run_action.py

```diff
@@ -495,6 +494,138 @@ def dismiss_change_requests(
+              comments(first: 1) {
+                nodes {
+                  id
+                  body
+                  author {
+                    login
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    """ % (
+        repo_owner,
+        repo_name,
+        pr_number,
+    )
+
+    response = requests.post(
+        "https://api.github.com/graphql",
+        json={"query": query},
+        headers={"Authorization": "Bearer " + github_token},
+        timeout=github_api_timeout,
+    )
+
+    if response.status_code == 200:
+        data = response.json()
+
+        # Iterate through review threads
+        for thread in data["data"]["repository"]["pullRequest"]["reviewThreads"][
+            "nodes"
+        ]:
+            for comment in thread["comments"]["nodes"]:
+                if (
+                    comment["id"]
+                    and thread["isResolved"] is False
+                    # this actor here is somehow different from `github-actions[bot]`
+                    # which we get through the Rest API
+                    and comment["author"]["login"] == "github-actions"
+                ):
+                    yield thread
+                    break
+    else:
+        print(
+            f"::error::getting unresolved conversation threads: {response.status_code}"
+        )
+        raise RuntimeError("Failed to get unresolved conversation threads.")
+
+
+def close_conversation(thread_id, github_token, github_api_timeout):
+    """Close a conversation thread using the GitHub GraphQL API"""
+    mutation = (
+        """
+    mutation {
+      resolveReviewThread(input: {threadId: "%s", clientMutationId: "github-actions"}) {
+        thread {
+          id
+        }
+      }
+    }
+    """
+        % thread_id
+    )
+
+    print(f"::debug::Closing conversation {thread_id}...")
+    response = requests.post(
+        "https://api.github.com/graphql",
+        json={"query": mutation},
+        headers={"Authorization": "Bearer " + github_token},
+        timeout=github_api_timeout,
+    )
+
+    if response.status_code == 200:
+        if "errors" in response.json():
+            msg = response.json()["errors"][0]["message"]
+            if "Resource not accessible by integration" in msg:
+                print(
+                    "::error::Closing conversations requires `contents: write` permission."
+                )
+            else:
+                print(f"::error::Closing conversation query failed: {msg}")
+        else:
+            print("Conversation closed successfully.")
+            return
+    else:
+        print(f"::error::GraphQL request failed: {response.status_code}")
+    print(
+        "::error:: Failed to close conversation. See log for details and "
+        "https://github.com/platisd/clang-tidy-pr-comments/blob/master/README.md for help"
+    )
+    raise RuntimeError("Failed to close conversation.")
```

Comment:
Can we please refactor this to decrease the "depth" of the function? E.g. it should look like this, with multiple return statements and no nested `if`s:

```python
if response.status_code != 200:
    print(f"::error::GraphQL request failed: {response.status_code}")
    # ...
    raise RuntimeError("Failed to close conversation.")
if "errors" not in response.json():
    print("Conversation closed successfully.")
    return
msg = response.json()["errors"][0]["message"]
# ....
```

Reply:
With the current flow we don't have to repeat the runtime error + print for the error cases in the `response.status_code == 200` branch.
Do you want me to pull the print+raise into a local function, do the reorder and call that function where needed?

Reply:
I am OK with both the local function and also with this small duplication. :+1: